API/Controllers/ClientsController.cs

```diff
@@ -0,0 +1,44 @@
+using API.DTOs;
+using API.Entities;
+using API.Services.Interfaces;
+using Asp.Versioning;
+using Microsoft.AspNetCore.Authorization;
+using Microsoft.AspNetCore.Cors;
+using Microsoft.AspNetCore.Http.HttpResults;
+using Microsoft.AspNetCore.JsonPatch;
+using Microsoft.AspNetCore.Mvc;
+
+namespace API.Controllers;
+
+[ApiController]
+[ApiVersion(1)]
+[EnableCors]
+[Route("api/v{version:apiVersion}/[controller]")]
+public class ClientsController(IOAuthClientService clientService) : Controller
+{
+    [HttpPatch("{id:int}/ratelimit")]
+    [Authorize(Roles = "admin")]
+    [EndpointSummary("Patches the ratelimit for a given client")]
+    public async Task<Results<BadRequest, NotFound<string>, Ok<OAuthClientDTO>>> PatchRatelimitAsync(int id, [FromBody] JsonPatchDocument<RateLimitOverrides> patchedOverrides)
+    {
+        OAuthClientDTO? currentClient = await clientService.GetAsync(id);
+
+        if (currentClient == null)
+        {
+            return TypedResults.NotFound("Client not found");
+        }
+
+        RateLimitOverrides overrides = currentClient.RateLimitOverrides;
+
+        if (patchedOverrides.Operations.Any(op => op.op != "replace"))
+        {
+            return TypedResults.BadRequest();
+        }
+
+        patchedOverrides.ApplyTo(overrides, ModelState);
+
+        // We know client is not null here due to the previous null check
+        OAuthClientDTO? client = await clientService.SetRatelimitOverridesAsync(id, overrides);
```

Comment:
We can just make the return type of `SetRatelimitOverridesAsync` non-nullable. We do this elsewhere in the codebase where a function will throw an error instead of returning null. At this point we are already assuming the client exists so it would make sense.

Reply:
I generally disagree with throwing exceptions for null items. The caller (controller in this case) should be able to gracefully handle these null results, rather than try-catching everything. We shouldn't modify service behavior assuming it will only be called in this fashion by the controller. It just so happens that the `SetRatelimitOverridesAsync` function also makes the exact same null check we make in the controller.

Reply:
Gotcha, that's fine with me then